File: patient-feedback-analytics-system-main/server/scripts/import_feedback.py
import csv
import os
from datetime import datetime
from pymongo import MongoClient

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
FEEDBACK_CSV = os.path.join(BASE_DIR, 'storage', 'feedback.csv')
USERS_CSV = os.path.join(BASE_DIR, 'storage', 'users.csv')
ENV_FILE = os.path.join(BASE_DIR, '.env')

# load MONGODB_URI from .env
MONGODB_URI = None
if os.path.exists(ENV_FILE):
    with open(ENV_FILE, 'r', encoding='utf8') as ef:
        for line in ef:
            if line.strip().startswith('MONGODB_URI='):
                MONGODB_URI = line.strip().split('=',1)[1]
                break

if not MONGODB_URI:
    raise SystemExit('MONGODB_URI not found in .env')

print('Using MongoDB URI:', MONGODB_URI[:40] + '...')

# build staffId -> name map from users.csv
staff_name = {}
if os.path.exists(USERS_CSV):
    with open(USERS_CSV, newline='', encoding='utf8') as f:
        reader = csv.DictReader(f)
        for r in reader:
            sid = (r.get('staffId') or '').strip()
            name = (r.get('name') or '').strip()
            if sid:
                staff_name[sid.upper()] = name

# connect MongoDB
client = MongoClient(MONGODB_URI)
db = client.get_default_database()
feedback_coll = db.get_collection('feedbacks')

# read feedback CSV
rows = []
with open(FEEDBACK_CSV, newline='', encoding='utf8') as f:
    reader = csv.DictReader(f)
    for r in reader:
        rows.append(r)

print('Total rows in feedback.csv:', len(rows))

insert_count = 0
for r in rows:
    created_str = (r.get('createdAt') or '').strip()
    if not created_str:
        created = datetime.utcnow()
    else:
        try:
            created = datetime.fromisoformat(created_str.replace('Z','+00:00'))
        except Exception:
            # try fallback
            created = datetime.utcnow()

    # avoid duplicate based on createdAt + staffId + text
    staffId = (r.get('doctorId') or '').strip() or None
    text = (r.get('text') or '').strip() or ''
    existing = feedback_coll.find_one({'createdAt': created, 'staffId': staffId, 'text': text})
    if existing:
        continue

    sentiment_raw = (r.get('sentiment') or '').strip().lower()
    sentiment = 'Neutral'
    if sentiment_raw.startswith('pos'):
        sentiment = 'Positive'
    elif sentiment_raw.startswith('neg'):
        sentiment = 'Negative'

    try:
        polarity = float(r.get('sentimentScore') or r.get('polarity') or 0)
    except:
        polarity = 0.0

    try:
        rating = int(r.get('rating') or 0)
    except:
        rating = 0

    doc = {
        'text': text,
        'department': 'OPD',
        'sentiment': sentiment,
        'polarityScore': polarity,
        'sentimentScore': polarity,
        'emotion': 'neutral',
        'category': 'General',
        'severityScore': max(1, min(10, (6 - rating) * 2)) if rating else 5,
        'keywords': [],
        'patientName': r.get('patientId') or 'Anon',
        'patientId': r.get('patientId') or 'Anon',
        'staffId': staffId,
        'staffName': staff_name.get((staffId or '').upper()) if staffId else None,
        'doctorRating': rating,
        'createdAt': created,
    }

    feedback_coll.insert_one(doc)
    insert_count += 1

print('Inserted', insert_count, 'new feedback documents')
client.close()

File: patient-feedback-analytics-system-main/server/scripts/import_feedback.js
import fs from "fs";
import path from "path";
import mongoose from "mongoose";
import Feedback from "../models/Feedback.js";

const BASE = path.resolve(process.cwd(), "server");
const USERS_CSV = path.join(BASE, "storage", "users.csv");
const FEEDBACK_CSV = path.join(BASE, "storage", "feedback.csv");
const ENV = path.join(BASE, ".env");

function readEnv(uriKey = "MONGODB_URI") {
  if (!fs.existsSync(ENV)) return null;
  const txt = fs.readFileSync(ENV, "utf8");
  for (const line of txt.split(/\r?\n/)) {
    const idx = line.indexOf(uriKey + "=");
    if (idx !== -1)
      return line.substring(idx + uriKey.length + 1).trim() || null;
  }
  return null;
}

function parseCSVLine(line) {
  const cols = [];
  let cur = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') {
        cur += '"';
        i++;
      } else inQuotes = !inQuotes;
      continue;
    }
    if (ch === "," && !inQuotes) {
      cols.push(cur);
      cur = "";
    } else {
      cur += ch;
    }
  }
  cols.push(cur);
  return cols;
}

async function main() {
  const MONGODB_URI = readEnv();
  if (!MONGODB_URI) {
    console.error("MONGODB_URI not found in .env");
    process.exit(1);
  }
  console.log("Connecting to MongoDB...");
  await mongoose.connect(MONGODB_URI, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
  });
  console.log("Connected");

  // load users to map staffId->name
  const staffName = {};
  if (fs.existsSync(USERS_CSV)) {
    const txt = fs.readFileSync(USERS_CSV, "utf8");
    const lines = txt.split(/\r?\n/).filter(Boolean);
    const hdr = lines
      .shift()
      .split(",")
      .map((h) => h.trim());
    for (const l of lines) {
      const cols = parseCSVLine(l);
      const row = {};
      hdr.forEach((h, i) => {
        row[h] = cols[i] || "";
      });
      if (row.staffId) staffName[row.staffId.toUpperCase()] = row.name || "";
    }
  }

  // read feedback csv
  if (!fs.existsSync(FEEDBACK_CSV)) {
    console.error("feedback.csv not found at", FEEDBACK_CSV);
    process.exit(1);
  }
  const ftxt = fs.readFileSync(FEEDBACK_CSV, "utf8");
  const lines = ftxt.split(/\r?\n/).filter(Boolean);
  const hdr = parseCSVLine(lines.shift()).map((h) => h.trim());

  const idx = (name) => hdr.indexOf(name);
  const iDoctor = idx("doctorId") >= 0 ? idx("doctorId") : idx("staffId");
  const iText = idx("text");
  const iRating = idx("rating");
  const iSentiment = idx("sentiment");
  const iScore =
    idx("sentimentScore") >= 0 ? idx("sentimentScore") : idx("polarity");
  const iCreated = idx("createdAt");
  const iPatient = idx("patientId");
  const iSpec = idx("specialization");

  let inserted = 0;
  for (const l of lines) {
    const cols = parseCSVLine(l);
    const doctorId = iDoctor >= 0 && cols[iDoctor] ? cols[iDoctor].trim() : "";
    const staffId = doctorId ? doctorId.trim() : null;
    const text = iText >= 0 && cols[iText] ? cols[iText].trim() : "";
    const rating = iRating >= 0 && cols[iRating] ? parseInt(cols[iRating]) : 0;
    const sentimentRaw =
      iSentiment >= 0 && cols[iSentiment]
        ? cols[iSentiment].trim().toLowerCase()
        : "";
    const score = iScore >= 0 && cols[iScore] ? parseFloat(cols[iScore]) : 0;
    const createdStr =
      iCreated >= 0 && cols[iCreated] ? cols[iCreated].trim() : "";
    const patientId =
      iPatient >= 0 && cols[iPatient] ? cols[iPatient].trim() : "Anon";
    const spec = iSpec >= 0 && cols[iSpec] ? cols[iSpec].trim() : "General";

    let created = new Date();
    if (createdStr) {
      created = new Date(createdStr);
      if (isNaN(created.getTime())) created = new Date();
    }

    // check duplicate
    const existing = await Feedback.findOne({
      createdAt: created,
      staffId: staffId,
      text: text,
    }).exec();
    if (existing) continue;

    let sentiment = "Neutral";
    if (sentimentRaw.startsWith("pos")) sentiment = "Positive";
    else if (sentimentRaw.startsWith("neg")) sentiment = "Negative";

    const severity = rating ? Math.max(1, Math.min(10, (6 - rating) * 2)) : 5;

    const doc = new Feedback({
      text: text,
      department: "OPD",
      sentiment: sentiment,
      polarityScore: score,
      sentimentScore: score,
      emotion: "neutral",
      category: "General",
      severityScore: severity,
      keywords: [],
      patientName: patientId,
      patientId: patientId,
      staffId: staffId,
      staffName: staffId ? staffName[staffId.toUpperCase()] || null : null,
      doctorRating: rating,
      createdAt: created,
    });

    await doc.save();
    inserted++;
  }

  console.log("Inserted", inserted, "feedback documents");
  await mongoose.disconnect();
  process.exit(0);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
